### PR TITLE
[stdlib] Conform fixed-width integer types to LosslessStringConvertible

### PR DESCRIPTION
**Pull request by xwu**
As approved in [SE-0089](https://github.com/apple/swift-evolution/blob/master/proposals/0089-rename-string-reflection-init.md), `Int` and its friends should conform to `LosslessStringConvertible`.

Since `init?(_:radix:)` is now implemented on `FixedWidthInteger`, this PR adopts the following design:

1. `FixedWidthInteger` now refines `LosslessStringConvertible`
2. The required initializer `init?(_:)` has a default implementation that calls `init?(_:radix:)`.

There's clearly a test that can be added to test this conformance. What's going to be even more critical is assuring source compatibility.

> Note: Currently, a method `foo(x: Int, y: Int = 42)` cannot satisfy a protocol requirement for `foo(x: Int)`; however, both can co-exist without the compiler complaining about ambiguity.

### CODE REVIEW

File: stdlib/public/core/IntegerParsing.swift [hunk 1]
```diff
@@ -155,4 +155,27 @@ extension FixedWidthInteger {
...
+  /// base 10 is not representable, the result is `nil`. For example, the
+  /// following conversions result in `nil`:
+  ///
+  ///     Int(" 100")                       // Includes whitespace
+  ///     Int("21-50")                      // Invalid format
+  ///     Int("ff6600")                     // Characters out of bounds
+  ///
+  /// - Parameter description: The ASCII representation of a number in base 10.
+  @_semantics("optimize.sil.specialize.generic.partial.never")
+  @inline(__always)
```
Comment (xwu) [inline stdlib/public/core/IntegerParsing.swift]: Are these appropriate and optimal uses of these internal attributes? I'm basing these off of the example of other initializers implemented in this file and `Integers.swift.gyb`.

**Comment by xwu** [comment]
I'm unable to test this locally, as the `ModelIO` overlay is refusing to build. Can we do a test?

**Comment by cardoso** [comment]
+1

This PR fixes: [SR-5558](https://bugs.swift.org/browse/SR-5558)

**Comment by huonw** [comment]
@swift-ci Please smoke test.

>  I'm unable to test this locally, as the ModelIO overlay is currently refusing to build. 

The work-around some of us are using is commenting out the contents of ModelIO.swift, since nothing uses it.

**Comment by huonw** [comment]
@swift-ci Please smoke test

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

File: stdlib/public/core/IntegerParsing.swift (hunk 1 above)
Comment (moiseev) [inline stdlib/public/core/IntegerParsing.swift]: /cc @swiftix for the usage of the `@_semantics` attribute.
`@inline(__always)` is fine.

Comment (moiseev) [inline stdlib/public/core/IntegerParsing.swift]: /cc @natecook1000 for the doc comment.

**Comment by moiseev** [comment]
@xwu Thanks!

> There's clearly a test that can be added to test this conformance.

Is there a reason you haven't implemented a test as part of this PR?

File: stdlib/public/core/IntegerParsing.swift (hunk 1 above)
Comment (natecook1000) [inline stdlib/public/core/IntegerParsing.swift]: Can you include an out of range example? e.g.:

    Int("10000000000000000000000000")               // Out of range

You could remove "in base 10" from the last paragraph, too.

File: stdlib/public/core/IntegerParsing.swift [hunk 2]
```diff
@@ -102,7 +102,7 @@ extension FixedWidthInteger {
   ///
   /// The string passed as `text` may begin with a plus or minus sign character
   /// (`+` or `-`), followed by one or more numeric digits (`0-9`) or letters
-  /// (`a-z` or `A-Z`). The string is case insensitive.
+  /// (`a-z` or `A-Z`). The string is case-insensitive.
```
Comment (natecook1000) [inline stdlib/public/core/IntegerParsing.swift]: This is correct as is. "The case-insensitive string…" vs "The string is case insensitive."

Comment (natecook1000) [inline stdlib/public/core/IntegerParsing.swift]: Okay, really the string is just a string. 😅 Can you change this to "The string parsing is case insensitive"?

**Review by natecook1000 — COMMENTED** [review]
Thanks @xwu — a couple minor requests.

**Comment by moiseev** [comment]
Somewhat related to https://github.com/apple/swift/pull/11157

File: stdlib/public/core/IntegerParsing.swift (hunk 2 above)
Comment (xwu) [inline stdlib/public/core/IntegerParsing.swift]: @natecook1000 You're right! My personal style is to hyphenate most compound adjectives (with the most common exception arising when the first word is an adverb that ends in "-ly"). However, _Chicago Manual_ rules (which, I believe, is or was at one point Apple's house style) calls for the spelling "case insensitive" when it follows the noun.

Also an astute point about what exactly is case-insensitive. I'd tweak it a little and write: "Parsing of the string is case insensitive."

**Comment by xwu** [comment]
@moiseev Yes, there was a very good reason: it was bedtime last night and I had fifteen hours of patient care duties today :) But now I've added one.

File: stdlib/public/core/IntegerParsing.swift (hunk 2 above)
Comment (natecook1000) [inline stdlib/public/core/IntegerParsing.swift]: Sounds great!

File: test/stdlib/Integers.swift.gyb [hunk 3]
```diff
@@ -364,6 +364,21 @@ tests.test("extendingOrTruncating") {
   expectEqual(129, UInt8(extendingOrTruncating: 129 + 1024 as UDWord))
 }
 
+tests.test("Parsing/LosslessStringConvertible") {
+  func _toArray<T: LosslessStringConvertible>(_ text: String) -> [T] {
+    return text.split(separator: " ").map { T(String($0)) }.flatMap { $0 }
+  }
+
+  let i: [Int] = _toArray("1 2 3")
+  expectEqual(i.count, 3)
```
Comment (moiseev) [inline test/stdlib/Integers.swift.gyb]: FYI: there exists 'expectEqualSequence` https://github.com/apple/swift/blob/master/stdlib/private/StdlibUnittest/StdlibUnittest.swift.gyb#L2326.

So one can write checks like these in one shot:

```swift
expectEqualSequence([1, 2, 3], _toArray("1 2 3"))
```

Comment (xwu) [inline test/stdlib/Integers.swift.gyb]: That's delightful.

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by xwu** [comment]
@moiseev Sorry, let's try this again; a minor change was necessary.

**Comment by moiseev** [comment]
@swift-ci Please test

**Comment by moiseev** [comment]
@swift-ci Please Test Source Compatibility

**Pull request merged**